### DISCUSSION TITLE
9-bit serial USART support?

### DISCUSSION THREAD

**Discussion by bstin** [micropython/micropython#15756] [SAMD]
According to the datasheet for the ATSAMD51 (available [here](https://ww1.microchip.com/downloads/aemDocuments/documents/MCU32/ProductDocuments/DataSheets/SAM-D5x-E5x-Family-Data-Sheet-DS60001507.pdf))

In Section 34., it says: 
* Supports Serial Frames with 5, 6, 7, 8 or **9 Data bits** and 1 or 2 Stop bits

I notice that there are specific ports of Micropython (like maybe the STM32?) that also seem to reference 9-bit serial support. Thus my question is does the Micropython port also support 9-bit for the SAMD51?

**Comment by ricksorensen** [comment]
Have you tried it? It looks like the configuration should handle setting 9-bit mode, but I have no ability to test it.

**Comment by robert-hh** [comment]
Sorry, no 9 bit support by Micropython fot SAMD51. The STM32 port is the only one which implements 9 bit serial, and it is possible to have it with RP2 PIO scripts.

**Reply by bstin** [reply]
Thanks for the confirmation. 

* I don't have the requisite C skills to add 9-bit support for SAMD51. Does the Micropython project accept bounties? I would be willing to contribute to a bounty if someone could add it. (Provided its technically possible to add it)

* I've looked at going the RPI Pico PIO route and that looks promising. However, one thing I don't understand. Is it possible to sub-class the existing UART peripheral to make it utilize a PIO-RX and PIO-TX State Machines? So create a 9-bit compatible state machines and plug them into the existing UART interface via a sub-class. I'm interested in just being able to use a standard UART style interface in my code and don't want to "re-invent the wheel" more than necessary.

**Reply by robert-hh** [reply]
Looking into the SAMD code, it may be possible. The setting is accepted, and 9 bit data is transferred, only with 9th bit set to 0 on writing, and reading ignores the 9th bit. Since UART provides a byte sized stream, it's the data transfer which has to be split in two byte sized items.

**Reply by bstin** [reply]
Excellent. Thank you @robert-hh - I have a SAMD device arriving today/tomorrow (a Seeed Wio Terminal), so I will test. 

The specific application I am using this for is not "true 9-bit serial", instead it uses the 9th bit as an address bit. Specifically it used to help establish the start of frames....so the first transmitted byte will have 9th bit set 1....then all subsequent bytes of same frame will have it set to 0. 

So given that, how can I "manually" set it to 1? I could then track this myself (I realize I could probably hack something together using parity bits, but frankly that seems fairly hacky and would prefer to have as much handled by hardware as possible).

**Reply by robert-hh** [reply]
I made a quick change to implement 9bit support. Not yet fully tested. The complication is, that for writing 9 bit data you have to use uart.write() with 2 bytes for each transmitted character, low order byte first. When reading with uart.read() you will get 2 bytes for each transmitted character, which you then have to assemble. As said, the Python API with uart.write() and uart,read() will still be byte-oriented. A build for a SEEED terminal is attached.

[firmware_SEEED_WIO.zip](https://github.com/user-attachments/files/16838757/firmware_SEEED_WIO.zip)

**Reply by bstin** [reply]
Wow! Thank you so much Robert!

In this case, I don't think the two-byte issue is much of a complication for me, as I can just default to sending (databyte, 0x00) and when I need to start a new frame issue a (databyte, 0x01). 

Can't wait until Wio comes in to test. Provided this works, I guess it would make it into next release?

**Reply by robert-hh** [reply]
> Provided this works, I guess it would make it into next release?

I first have to test and make a PR. Since the next release is expected soon, it might have to wait for the following release.

**Reply by robert-hh** [reply]
With the existing firmware you could send the first byte as 8 bit, 2 stop bits, and the remaining ones as 9 bit, 1 stop bit. That should have the 9th bit at the first byte at 1, and at 0 for the following bit. You can switch mode with uart.init(). Only you have to wait until the first byte has been sent.

**Reply by bstin** [reply]
Just as an FYI. I notice the SAMD21 also supports 9-bit ([datasheet](https://www.mouser.com/datasheet/2/744/ETA_USA_102010328___Xiao-3305560.pdf) - section 26.2). This includes devices like the Seeed XIAO SAMD21 + various Adafruit / Sparkfun / etc. Just mentioning this as I think your PR could apply to both SAMD51/21.

**Reply by robert-hh** [reply]
The UART code is mostly identical for SAMD21 and SAMD51.

**Reply by robert-hh** [reply]
Tested with a SAMD21 and it works as well.

**Reply by bidrohini** [reply]
I'm afraid that modification of the MicroPython UART driver is the only way. Because micropython  primarily supports 7-bit or 8-bit data frames.

**Reply by robert-hh** [reply]
I made a PR #15787 for UART 9 bit support.

**Reply by bstin** [reply]
Many thanks Robert!

Unrelated, but have you heard any recent news about Seeed / RTL enabling WiFi for Wio Terminal? Would be great to have WiFi on this as well....

**Reply by robert-hh** [reply]
>  Would be great to have WiFi on this as well....

Sure. Only someone has to implement the interface between MicroPython and the RTL firmware, which is a larg piece of work. Seeed and RTL will not do it, I assume. AFAIK, they offer a Arduino library, which could be used as a starting point.
